### ISSUE TITLE
--with-ugni=no is not honoured if ugni libraries installed

### ISSUE THREAD

**Issue by chrissamuel** [openucx/ucx#3135]
Hi there,

I am trying to help someone get OMPI and UCX going on some x86-64 Infiniband connected nodes that use an image based on Cray eslogin nodes but with OpenSuSE as the base rather than SLES12.  Consequently there are various Cray libraries installed and it does not appear to be able to disable the use of ugni with them present.

Using the 1.5.0RC1 tarball I've been able to fix this by changing this line in Configure:

`  withval=$with_ugni; with_ugni=forced`

to

` withval=$with_ugni; with_ugni=no`

As otherwise the next step tests for it being no and sees it is forced instead.

This is generated by autoconf but I'm guessing this is due to the logic here:

https://github.com/openucx/ucx/blob/v1.5.0-rc1/config/m4/cray_ugni.m4

Any ideas?

All the best,
Chris

**Comment by dmjacobsen** [comment]
After Chris posted this in our chat, I started looking at the master branch (where config/m4/cray_ugni.m4 has been moved to src/uct/ugni/configure.m4)

In any case I think this patch fixes it:

```
diff --git a/src/uct/ugni/configure.m4 b/src/uct/ugni/configure.m4
index b6efa6e8..b10831de 100644
--- a/src/uct/ugni/configure.m4
+++ b/src/uct/ugni/configure.m4
@@ -9,8 +9,8 @@ cray_ugni_supported=no
 AC_ARG_WITH([ugni],
         [AC_HELP_STRING([--with-ugni(=DIR)],
             [Build Cray UGNI support, adding DIR/include, DIR/lib, and DIR/lib64 to the search path for headers and libraries])],
-        [with_ugni=forced],
-        [with_ugni=yes])
+       [],
+       [with_ugni=default])

 AS_IF([test "x$with_ugni" != "xno"],
         [PKG_CHECK_MODULES([CRAY_UGNI], [cray-ugni cray-pmi],
@@ -18,7 +18,7 @@ AS_IF([test "x$with_ugni" != "xno"],
                             cray_ugni_supported=yes
                             AC_DEFINE([HAVE_TL_UGNI], [1],
                                       [Define if UGNI transport exists.])],
-                           [AS_IF([test "x$with_ugni" == "xforced"],
+                           [AS_IF([test "x$with_ugni" != "xdefault"],
                                   [AC_MSG_WARN([UGNI support was requested but cray-ugni and cray-pmi packages can't be found])
                                    AC_MSG_ERROR([Cannot continue])],[])]
                            )])
```

The current logic disregards anything the user provides because it sets with_ugni to forced if the user provides any argument.  The supplied patch reverses the logic by explicitly setting a default value of default and checking if the user asked explicitly later (for any value other than "no") by checking it isn't the default.

The suggested patch should also fix the similarly rare case where the user wants a specific directory searched for ugni libraries.

**Comment by miked-mellanox** [comment]
maybe you need try with ```./configure --without-ugni ...```

**Comment by dmjacobsen** [comment]
--without-ugni is non-functional for the same reason as above -- the AC_ARG_WITH _action-if-given_ argument overrides anything the user submits.  Also, we tried it =)

**Comment by dmjacobsen** [comment]
PR 3136 addresses this.

**Comment by shamisp** [comment]
@manjugv @hjelmn can you please take a looks since it is ugni related.

**Comment by dmjacobsen** [comment]
Added PR 3144 to address this in 1.5.x

**Comment by shamisp** [comment]
Fixed in #3136

**Comment by chrissamuel** [comment]
Thank you so much everyone!
